### PR TITLE
ingress facade

### PR DESCRIPTION
**Pull request by sed-i**
Encapsulating all ingress related logic into a Consumer-like library.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -17,13 +17,53 @@
...
+from ops.framework import EventSource, EventBase
+
 
 PROMETHEUS_CONFIG = "/etc/prometheus/prometheus.yml"
 RULES_DIR = "/etc/prometheus/rules"
 
 logger = logging.getLogger(__name__)
 
 
+class IngressConsumer(ConsumerBase):
```
Comment (balbirthomas) [inline src/charm.py:30]: Good idea. But why not create a PR against the Ingress charm. Also may be a good idea to factor this out into a separate file `src/ingress.py` and if required any unit tests for it `tests/test_ingress.py`.

Comment (mmanciop) [reply src/charm.py:30]: I'd be in favor of proposing this as a PR for the upstream repo

File: tox.ini [hunk 2]
```diff
@@ -29,7 +29,7 @@ deps =
     flake8
 commands =
     flake8 {toxinidir}/src {toxinidir}/tests {toxinidir}/lib/charms/prometheus_k8s
-    black --check --diff --target-version=py38 {toxinidir}/src {toxinidir}/tests {toxinidir}/lib/charms/prometheus_k8s
+    black --line-length=99 --check --diff --target-version=py38 {toxinidir}/src {toxinidir}/tests {toxinidir}/lib/charms/prometheus_k8s
```
[the remark this answers is not in the document]
Comment (sed-i) [reply tox.ini:32]: We use 99 throughout, for example
https://github.com/canonical/alertmanager-operator/blob/6e785756a98e60d005fc4ec5a561cc408d5a8668/tox.ini#L36
but it sounds like you intentionally omitted it from both flak8 and black... :)

[the remark this answers is not in the document]
Comment (balbirthomas) [reply tox.ini:32]: You seem to be referring to yourself as "we" :-) . I never remember this being a discussed.

File: src/charm.py (hunk 1 above)
Comment (sed-i) [reply src/charm.py:30]: @mthaddon @jnsgruk what do you think about adding this as a library to [charm-k8s-ingress](https://launchpad.net/charm-k8s-ingress)? ([diff ref](https://github.com/canonical/prometheus-operator/pull/97/commits/5ccc81b5736e34e8ea84afa810e6a3d45fbceda0))

File: tox.ini (hunk 2 above)
[the remark this answers is not in the document]
Comment (sed-i) [reply tox.ini:32]: prometheus had 99 before tox migration
https://github.com/canonical/prometheus-operator/blob/34c0dca535c422029e50ba88c3897364031d42ff/.flake8#L2

other charms as well.

[the remark this answers is not in the document]
Comment (balbirthomas) [reply tox.ini:32]: Which was not default behaviour and I found myself surprised when creating tox ini that black reformatted my code. Then I realized this. In standup we had agreed to stick with black defaults.

File: lib/charms/prometheus_k8s/v0/prometheus.py [hunk 3]
```diff
@@ -462,13 +458,9 @@ def alerts(self):
             if not relation.units:
                 continue
 
-            alert_rules = json.loads(
-                relation.data[relation.app].get("alert_rules", "{}")
-            )
```
Comment (jnsgruk) [inline lib/charms/prometheus_k8s/v0/prometheus.py:467]: There is a lot of this which is clearly a `black` thing - can you seperate the work you're doing from the `black` formatting to make the diff a bit more readable?

File: src/charm.py (hunk 1 above)
Comment (jnsgruk) [reply src/charm.py:30]: What are we proposing to add? The `ingress-changed` event? If so, I don't see a strong reason why that couldn't be added to @mthaddon's upstream library.

File: lib/charms/prometheus_k8s/v0/prometheus.py (hunk 3 above)
Comment (sed-i) [reply lib/charms/prometheus_k8s/v0/prometheus.py:467]: it is already divided into two separate commits

File: tox.ini (hunk 2 above)
[the remark this answers is not in the document]
Comment (balbirthomas) [reply tox.ini:32]: Ok. Let us merge the reformatting commit.

File: src/charm.py (hunk 1 above)
Comment (sed-i) [reply src/charm.py:30]: Essentially replace
`self.ingress = IngressRequires(...)`
with
`self.ingress_lib = IngressConsumer(...)`

to be more precise:
1. Add a consumer library - new class that inherits from `ops.relation.ConsumerBase`
2. Delegate all relevant event observation to the new class
3. Expect users to observe a (custom) event coming from the new lib, `self.ingress_lib.ingress_changed`

Whoever currently uses `IngressRequires` directly could still continue and do it.

File: lib/charms/prometheus_k8s/v0/prometheus.py (hunk 3 above)
Comment (sed-i) [reply lib/charms/prometheus_k8s/v0/prometheus.py:467]: let me drop that linting commit for now (gh action may fail)

**Comment by balbirthomas** [comment]
Closing this, as it is no longer relevant.

**Pull request closed without merging**